### PR TITLE
CC-2049, CC-2060: Rewrite backups to operate on a single tar stream.

### PR DESCRIPTION
**Pull request by iancmcc**
Backwards compatible with the previous backups.

This fixes tickets:
https://jira.zenoss.com/browse/CC-2049
https://jira.zenoss.com/browse/CC-2060

### CODE REVIEW

File: dfs/backup.go [hunk 1]
```diff
@@ -34,30 +37,18 @@ const (
...
-	if _, err := buffer.WriteTo(tarfile); err != nil {
-		glog.Errorf("Could not write backup metadata: %s", err)
+
+	// A set of pipes that output tar streams
+	var tarpipes []pipe.Pipe
+
+	// write the backup metadata. This is too tiny to parallelize
+	metadataBuffer := &bytes.Buffer{}
+	if err := dfs.writeBackupMetadata(data, metadataBuffer); err != nil {
+		glog.V(2).Infof("Unable to write backup metadata: %s", err)
```
Comment (smousa) [inline dfs/backup.go]: should this be an errorf?

File: dfs/backup.go [hunk 2]
```diff
@@ -82,7 +74,7 @@ func (dfs *DistributedFilesystem) Backup(data BackupInfo, w io.Writer) error {
 			return err
 		}
 		// load the images from this snapshot
-		glog.Infof("Preparing images for tenant %s", info.TenantID)
+		glog.V(2).Infof("Preparing images for tenant %s", info.TenantID)
```
Comment (smousa) [inline dfs/backup.go]: I don't know if V(2) is the way to go for this.

File: dfs/backup.go (hunk 1 above)
Comment (iancmcc) [inline dfs/backup.go]: That's a good question, and I don't know the right answer. The reason I made it not an errorf is because it ends up being double-logged then, once here, and once when the error gets returned up the stack. I think the right approach is debug log it at this low level, and error log it at the high level.

File: dfs/dfs.go [hunk 3]
```diff
@@ -54,7 +54,7 @@ type DFS interface {
 	// Backup saves and exports the current state of the system
 	Backup(info BackupInfo, w io.Writer) error
 	// Restore restores the system to the state of the backup
-	Restore(r io.Reader) (*BackupInfo, error)
+	Restore(r io.Reader, backupInfo *BackupInfo) error
```
Comment (smousa) [inline dfs/dfs.go]: Why did you change the signature here?

File: dfs/restore.go [hunk 4]
```diff
@@ -18,89 +18,82 @@ import (
...
-var ErrRestoreNoInfo = errors.New("backup is missing metadata")
+var (
+	ErrRestoreNoInfo        = errors.New("backup is missing metadata")
+	ErrInvalidBackupVersion = errors.New("backup has an invalid version")
+)
+
+// ExtractBackupInfo extracts the backup metadata from a tarball on disk in as
+// cheaply a manner as possible. The serialized BackupInfo is stored at the
+// front of the tarball to facilitate this.
+func ExtractBackupInfo(filename string) (*BackupInfo, error) {
```
Comment (smousa) [inline dfs/restore.go]: I don't think this makes sense to be in this file.  I think this should probably go into backupinfo.go

File: dfs/dfs.go (hunk 3 above)
Comment (iancmcc) [inline dfs/dfs.go]: Because we can extract the BackupInfo most cheaply from the file, rather than from the stream. Otherwise we have to recreate the stream after getting out the info. It seemed silly to pull it out to check the version and then pull it out again inside the restore.

File: dfs/restore.go (hunk 4 above)
Comment (iancmcc) [inline dfs/restore.go]: Done.

File: dfs/backup.go (hunk 2 above)
Comment (iancmcc) [inline dfs/backup.go]: The backup logging was too chatty, so I took stuff down. I can bring stuff back up, but this (along with others) seemed like a message you would only ever care about if something went wrong, hence debug.

File: dfs/backup.go [hunk 5]
```diff
@@ -105,41 +97,68 @@ func (dfs *DistributedFilesystem) Backup(data BackupInfo, w io.Writer) error {
...
+}
+
+// dockerSavePipe streams the tar archive output by Docker to pipe's stdout
+func (dfs *DistributedFilesystem) dockerSavePipe(images ...string) pipe.Pipe {
+	return pipe.TaskFunc(func(s *pipe.State) error {
+		return dfs.docker.SaveImages(images, s.Stdout)
+	})
+}
+
+func (dfs *DistributedFilesystem) snapshotSavePipe(vol volume.Volume, label string) pipe.Pipe {
```
Comment (smousa) [inline dfs/backup.go]: missing comment

File: dfs/restore.go [hunk 6]
```diff
@@ -109,6 +85,188 @@ func (dfs *DistributedFilesystem) Restore(r io.Reader) (*BackupInfo, error) {
...
+	wg.Wait()
+	// Return any errors, if produced
+	if len(errs) > 0 {
+		return errs[0]
+	}
+	// Everything's ok, so finalize by pushing images to the registry if necessary
+	return dfs.loadRegistry(registryImages)
+}
+
+func (dfs *DistributedFilesystem) restoreVolume(tenant, label string, tarfile io.Reader) ([]string, error) {
```
Comment (smousa) [inline dfs/restore.go]: missing comment
